### ISSUE TITLE
lineWrapping: cursor should skip last whitespace when moving right + incorrect char width

### ISSUE THREAD

**Issue by dwelle** [codemirror/CodeMirror#4841]
Refer to this [example](https://jsfiddle.net/ukyd792u/3/).

For `textarea` inputStyle, when a line wraps on whitespace, and you move cursor from the last word-character on line `0` _right_, it should move the cursor to line `1`. CM currently advances cursor _after_ the whitespace (last character on line `0`).

This results in inconsistent and unexpected behavior (all text editors wrap the cursor to the next line as described above). You can see the _standard_ behavior when you switch to `contenteditable` inputStyle in that example.

Inconsistency arises when you move the cursor (refer to the example) manually via keyboard/command right which moves it _after_ the whitespace still on line `0`, while if you move the cursor programmatically (`cm.setCursor{ line: 0, ch: 40 }`), it correctly puts the cursor at the start of line `1`.

From this incorrect behavior arises related bug which we'd not have to solve otherwise: the last whitespace has variable width depending on the editor size, and may end up just 1px wide:

![image](https://user-images.githubusercontent.com/5153846/27985187-f0f04b36-63e6-11e7-82d7-4a40fb205582.png)

**Comment by dwelle** [comment]
And now it becomes obvious why `goLineLeft` doesn't work correctly (https://github.com/codemirror/CodeMirror/issues/4840) when (visually) at the end of line `0`, because the `charCoords()` checks char 40 which _is_ at line `1` (but the `cursorCoords()` checks cursor which is at line `0`).

**Comment by dwelle** [comment]
Solution seems simple: Set position's `sticky: "after"` when moving horizontally (unless there's something that'll break due to this change, but I think it shouldn't).

You can check a proposed fix [here](https://github.com/dwelle/CodeMirror/commit/dd61fecf5b9ed01383687794ff8bafeadd997af6) - it also contains changes to `goLineRight` to make the behavior consistent, and also fixes related issue: https://github.com/codemirror/CodeMirror/issues/4840

**Comment by marijnh** [comment]
Simply setting `sticky` to `"after"` is not the solution -- it's set to `"before"` for a reason, namely that in principle, the cursor should stick to the character it just moved across when moving right. When there's no whitespace around the line-wrapping point, it _should_ appear at the end of the current line before moving to the next line.

@adrianheine Do you think it would be safe to set the stickiness to `"after"` in `moveLogically`/`moveVisually` whenever a whitespace character was skipped?

**Comment by dwelle** [comment]
> When there's no whitespace around the line-wrapping point, it should appear at the end of the current line before moving to the next line.

Not necessarily. E.g. Atom, Brackets, or notepad++ doesn't do that -- they don't allow you to move the cursor _after_ the last character on a wrapped line if there's no space after it, and instead move it to next line:

![image](https://user-images.githubusercontent.com/5153846/28009806-1037b0d4-655d-11e7-8bdf-06a3c9b28d1d.png)

Sublime text _does_ allow that, __but__ to counterbalance it, it doesn't allow that from the left-to-right movement on the next line.

So, we should decide on a behavior (I'd go with the one used in most editors, i.e. the first one), and stick to it.

Problem with current behavior is that it's not consistent. It allows both behaviors which is not consistent and thus results in incorrect behavior (as demonstrated e.g. when setting cursor programmatically).

**Comment by Niharika29** [comment]
I'm facing a similar issue and wondering if there's a config option to allow CodeMirror to leave the cursor movement up to the browser?

**Comment by dwelle** [comment]
@Niharika29 no... there are 2 different inputStyles (`contenteditable` & `textarea`, one used on mobile and the other on desktop, simply put) and if we let browsers do their own thing on each (which is not possibly anyway, at least for `textarea` inputStyle, due to how CM works), we would end up with potentially wildly inconsistent behavior (WTBS, currently there still are inconsistencies stemming up from some of the parts of `contenteditable` behavior being left up to the browser).

In any case, even native behavior **varies across browsers**. If you [try this fiddle](https://jsfiddle.net/cffcoo7k/), chrome handles cursor movement "correctly", while FF has the same inconsistent behavior as CM currently does. So deferring to native browser behavior wouldn't solve this problem.

I'm gonna issue a PR addressing this, but we need to settle on a solution (there are 2 ways to do this, both correct). Also it'll be a breaking change not to take lightly (unless we put it under an option, but that could complicate the codebase considerably, and I think CM should handle cursor movement correctly by default).
